IllegalStateException: Failed to recover 'Broker-0-StreamProcessor-2'

**Describe the bug**
After redeploying the cluster, it is falling into an IllegalStateException state. The broker is still up, but rejecting all gRPC requests with `INTERNAL` error. It is happening on my beta environment, but not on alpha environment. My dev cluster was fine till few hours back, but now I am seeing same gRPC `INTERNAL` errors on dev as well. Alpha is still fine by the way.

**Expected behavior**
The cluster should accept all gRPC requests.

<details><summary>Full Stacktrace</summary>
 <p>

```
2020-02-25 13:39:27.859 [Broker-0-StreamProcessor-2] [Broker-1-zb-actors-0] DEBUG io.zeebe.logstreams - Recovering state of partition 2 from snapshot
2020-02-25 13:39:28.608 [Broker-0-StreamProcessor-2] [Broker-1-zb-actors-0] ERROR io.zeebe.util.actor - Actor failed in phase 'STARTED'. Continue with next job.
java.lang.IllegalStateException: Expected to find event with the snapshot position 64424590432 in log stream, but nothing was found. Failed to recover 'Broker-0-StreamProcessor-2'.
	at io.zeebe.engine.processor.StreamProcessor.recoverFromSnapshot(StreamProcessor.java:209) ~[zeebe-workflow-engine-0.22.1.jar:0.22.1]
	at io.zeebe.engine.processor.StreamProcessor.onActorStarted(StreamProcessor.java:121) ~[zeebe-workflow-engine-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:73) ~[zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:115) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:107) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:91) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:195) [zeebe-util-0.22.1.jar:0.22.1]
2020-02-25 13:39:28.608 [Broker-1-ZeebePartition-2] [Broker-1-zb-actors-0] ERROR io.zeebe.broker.system - Unexpected error on stream processor installation!
java.lang.IllegalStateException: Expected to find event with the snapshot position 64424590432 in log stream, but nothing was found. Failed to recover 'Broker-0-StreamProcessor-2'.
	at io.zeebe.engine.processor.StreamProcessor.recoverFromSnapshot(StreamProcessor.java:209) ~[zeebe-workflow-engine-0.22.1.jar:0.22.1]
	at io.zeebe.engine.processor.StreamProcessor.onActorStarted(StreamProcessor.java:121) ~[zeebe-workflow-engine-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:73) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:115) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:107) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:91) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:195) [zeebe-util-0.22.1.jar:0.22.1]
2020-02-25 13:39:28.609 [Broker-1-ZeebePartition-2] [Broker-1-zb-actors-0] ERROR io.zeebe.broker.system - Failed to install leader partition 2
java.lang.IllegalStateException: Expected to find event with the snapshot position 64424590432 in log stream, but nothing was found. Failed to recover 'Broker-0-StreamProcessor-2'.
	at io.zeebe.engine.processor.StreamProcessor.recoverFromSnapshot(StreamProcessor.java:209) ~[zeebe-workflow-engine-0.22.1.jar:0.22.1]
	at io.zeebe.engine.processor.StreamProcessor.onActorStarted(StreamProcessor.java:121) ~[zeebe-workflow-engine-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:73) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:115) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:107) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:91) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:195) [zeebe-util-0.22.1.jar:0.22.1]
2020-02-25 13:39:28.609 [Broker-0-LogStream-2] [Broker-1-zb-actors-0] INFO  io.zeebe.logstreams - Close appender for log stream raft-partition-partition-2
2020-02-25 13:39:28.609 [Broker-0-StreamProcessor-2] [Broker-1-zb-actors-0] DEBUG io.zeebe.logstreams - Closed stream processor controller Broker-0-StreamProcessor-2.
2020-02-25 13:39:28.614 [Broker-0-StreamProcessor-3] [Broker-1-zb-actors-0] DEBUG io.zeebe.logstreams - Recovering state of partition 3 from snapshot
2020-02-25 13:39:29.185 [Broker-0-StreamProcessor-3] [Broker-1-zb-actors-0] ERROR io.zeebe.util.actor - Actor failed in phase 'STARTED'. Continue with next job.
java.lang.IllegalStateException: Expected to find event with the snapshot position 60130044736 in log stream, but nothing was found. Failed to recover 'Broker-0-StreamProcessor-3'.
	at io.zeebe.engine.processor.StreamProcessor.recoverFromSnapshot(StreamProcessor.java:209) ~[zeebe-workflow-engine-0.22.1.jar:0.22.1]
	at io.zeebe.engine.processor.StreamProcessor.onActorStarted(StreamProcessor.java:121) ~[zeebe-workflow-engine-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:73) ~[zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:115) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:107) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:91) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:195) [zeebe-util-0.22.1.jar:0.22.1]
2020-02-25 13:39:29.186 [Broker-1-ZeebePartition-3] [Broker-1-zb-actors-0] ERROR io.zeebe.broker.system - Unexpected error on stream processor installation!
java.lang.IllegalStateException: Expected to find event with the snapshot position 60130044736 in log stream, but nothing was found. Failed to recover 'Broker-0-StreamProcessor-3'.
	at io.zeebe.engine.processor.StreamProcessor.recoverFromSnapshot(StreamProcessor.java:209) ~[zeebe-workflow-engine-0.22.1.jar:0.22.1]
	at io.zeebe.engine.processor.StreamProcessor.onActorStarted(StreamProcessor.java:121) ~[zeebe-workflow-engine-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:73) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:115) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:107) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:91) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:195) [zeebe-util-0.22.1.jar:0.22.1]
2020-02-25 13:39:29.186 [Broker-1-ZeebePartition-3] [Broker-1-zb-actors-0] ERROR io.zeebe.broker.system - Failed to install leader partition 3
java.lang.IllegalStateException: Expected to find event with the snapshot position 60130044736 in log stream, but nothing was found. Failed to recover 'Broker-0-StreamProcessor-3'.
	at io.zeebe.engine.processor.StreamProcessor.recoverFromSnapshot(StreamProcessor.java:209) ~[zeebe-workflow-engine-0.22.1.jar:0.22.1]
	at io.zeebe.engine.processor.StreamProcessor.onActorStarted(StreamProcessor.java:121) ~[zeebe-workflow-engine-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:73) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:115) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:107) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:91) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:195) [zeebe-util-0.22.1.jar:0.22.1]
2020-02-25 13:39:29.187 [Broker-0-LogStream-3] [Broker-1-zb-actors-0] INFO  io.zeebe.logstreams - Close appender for log stream raft-partition-partition-3
2020-02-25 13:39:29.187 [Broker-0-StreamProcessor-3] [Broker-1-zb-actors-0] DEBUG io.zeebe.logstreams - Closed stream processor controller Broker-0-StreamProcessor-3.
2020-02-25 13:39:29.187 [raft-partition-partition-2-write-buffer] [Broker-1-zb-actors-0] DEBUG io.zeebe.dispatcher - Dispatcher closed
2020-02-25 13:39:29.187 [raft-partition-partition-3-write-buffer] [Broker-1-zb-actors-0] DEBUG io.zeebe.dispatcher - Dispatcher closed
2020-02-25 13:39:32.638 [GatewayTopologyManager] [Broker-1-zb-actors-0] DEBUG io.zeebe.gateway - Received broker was removed BrokerInfo{nodeId=2, partitionsCount=3, clusterSize=3, replicationFactor=3, partitionRoles={1=FOLLOWER, 2=FOLLOWER, 3=FOLLOWER}, partitionLeaderTerms={}}.
2020-02-25 13:39:32.639 [Broker-1-TopologyManager] [Broker-1-zb-actors-0] DEBUG io.zeebe.broker.clustering - Received member removed BrokerInfo{nodeId=2, partitionsCount=3, clusterSize=3, replicationFactor=3, partitionRoles={1=FOLLOWER, 2=FOLLOWER, 3=FOLLOWER}, partitionLeaderTerms={}} 
2020-02-25 13:40:03.405 [GatewayTopologyManager] [Broker-1-zb-actors-0] DEBUG io.zeebe.gateway - Received metadata change from Broker 2, partitions {} and terms {}.
2020-02-25 13:40:03.405 [Broker-1-TopologyManager] [Broker-1-zb-actors-0] DEBUG io.zeebe.broker.clustering - Received metadata change for 2, partitions {} terms {}
2020-02-25 13:40:04.404 [GatewayTopologyManager] [Broker-1-zb-actors-0] DEBUG io.zeebe.gateway - Received metadata change from Broker 2, partitions {3=FOLLOWER} and terms {}.
2020-02-25 13:40:04.405 [Broker-1-TopologyManager] [Broker-1-zb-actors-0] DEBUG io.zeebe.broker.clustering - Received metadata change for 2, partitions {3=FOLLOWER} terms {}
2020-02-25 13:40:05.154 [GatewayTopologyManager] [Broker-1-zb-actors-0] DEBUG io.zeebe.gateway - Received metadata change from Broker 2, partitions {2=FOLLOWER, 3=FOLLOWER} and terms {}.
2020-02-25 13:40:05.155 [Broker-1-TopologyManager] [Broker-1-zb-actors-0] DEBUG io.zeebe.broker.clustering - Received metadata change for 2, partitions {2=FOLLOWER, 3=FOLLOWER} terms {}
2020-02-25 13:40:05.904 [GatewayTopologyManager] [Broker-1-zb-actors-0] DEBUG io.zeebe.gateway - Received metadata change from Broker 2, partitions {1=FOLLOWER, 2=FOLLOWER, 3=FOLLOWER} and terms {}.
2020-02-25 13:40:05.905 [Broker-1-TopologyManager] [Broker-1-zb-actors-0] DEBUG io.zeebe.broker.clustering - Received metadata change for 2, partitions {1=FOLLOWER, 2=FOLLOWER, 3=FOLLOWER} terms {}
2020-02-25 13:43:49.394 [GatewayTopologyManager] [Broker-1-zb-actors-0] DEBUG io.zeebe.gateway - Received REACHABILITY_CHANGED for broker 0, do nothing.
2020-02-25 13:43:49.394 [Broker-1-TopologyManager] [Broker-1-zb-actors-0] DEBUG io.zeebe.broker.clustering - Received REACHABILITY_CHANGED from member 0, was not handled.
2020-02-25 13:43:59.638 [GatewayTopologyManager] [Broker-1-zb-actors-0] DEBUG io.zeebe.gateway - Received broker was removed BrokerInfo{nodeId=0, partitionsCount=3, clusterSize=3, replicationFactor=3, partitionRoles={1=FOLLOWER, 2=FOLLOWER, 3=FOLLOWER}, partitionLeaderTerms={}}.
2020-02-25 13:43:59.639 [Broker-1-TopologyManager] [Broker-1-zb-actors-0] DEBUG io.zeebe.broker.clustering - Received member removed BrokerInfo{nodeId=0, partitionsCount=3, clusterSize=3, replicationFactor=3, partitionRoles={1=FOLLOWER, 2=FOLLOWER, 3=FOLLOWER}, partitionLeaderTerms={}} 
2020-02-25 13:44:25.471 [Broker-1-ZeebePartition-1] [Broker-1-zb-actors-0] DEBUG io.zeebe.broker.system - Removing leader partition services for partition PartitionId{id=1, group=raft-partition}
2020-02-25 13:44:25.471 [Broker-1-ZeebePartition-1] [Broker-1-zb-actors-0] DEBUG io.zeebe.broker.system - Closing Broker-0-StreamProcessor-1
2020-02-25 13:44:25.471 [Broker-1-ZeebePartition-1] [Broker-1-zb-actors-0] DEBUG io.zeebe.broker.system - Partition role transitioning from LEADER to FOLLOWER
2020-02-25 13:44:25.471 [Broker-1-ZeebePartition-1] [Broker-1-zb-actors-0] DEBUG io.zeebe.broker.system - Closed Broker-0-StreamProcessor-1 successfully
2020-02-25 13:44:25.471 [Broker-1-ZeebePartition-1] [Broker-1-zb-actors-0] DEBUG io.zeebe.broker.system - Closing io.zeebe.broker.logstreams.LogStreamDeletionService
2020-02-25 13:44:25.472 [Broker-1-ZeebePartition-1] [Broker-1-zb-actors-0] DEBUG io.zeebe.broker.system - Closed io.zeebe.broker.logstreams.LogStreamDeletionService successfully
2020-02-25 13:44:25.473 [Broker-1-ZeebePartition-1] [Broker-1-zb-actors-0] DEBUG io.zeebe.logstreams.snapshot - Closed database from '/usr/local/zeebe/data/raft-partition/partitions/1/runtime'.
2020-02-25 13:44:25.473 [Broker-1-ZeebePartition-1] [Broker-1-zb-actors-0] DEBUG io.zeebe.broker.system - Installing follower partition service for partition PartitionId{id=1, group=raft-partition}
2020-02-25 13:44:25.479 [Broker-1-ZeebePartition-1] [Broker-1-zb-actors-0] DEBUG io.zeebe.logstreams.snapshot - Available snapshots: [SnapshotImpl{position=68719979328, path=/usr/local/zeebe/data/raft-partition/partitions/1/snapshots/314340-19-1581687776238-68719979328}, SnapshotImpl{position=68719943328, path=/usr/local/zeebe/data/raft-partition/partitions/1/snapshots/314201-19-1581686876225-68719943328}, SnapshotImpl{position=68719907328, path=/usr/local/zeebe/data/raft-partition/partitions/1/snapshots/314064-19-1581685976198-68719907328}]
2020-02-25 13:44:25.516 [Broker-1-ZeebePartition-1] [Broker-1-zb-actors-0] DEBUG io.zeebe.logstreams.snapshot - Opened database from '/usr/local/zeebe/data/raft-partition/partitions/1/runtime'.
2020-02-25 13:44:25.517 [Broker-1-ZeebePartition-1] [Broker-1-zb-actors-0] DEBUG io.zeebe.logstreams.snapshot - Recovered state from snapshot 'SnapshotImpl{position=68719979328, path=/usr/local/zeebe/data/raft-partition/partitions/1/snapshots/314340-19-1581687776238-68719979328}'
2020-02-25 13:44:26.523 [GatewayTopologyManager] [Broker-1-zb-actors-0] DEBUG io.zeebe.gateway - Received metadata change from Broker 0, partitions {} and terms {}.
2020-02-25 13:44:26.523 [Broker-1-TopologyManager] [Broker-1-zb-actors-0] DEBUG io.zeebe.broker.clustering - Received metadata change for 0, partitions {} terms {}
2020-02-25 13:44:26.773 [GatewayTopologyManager] [Broker-1-zb-actors-0] DEBUG io.zeebe.gateway - Received metadata change from Broker 0, partitions {3=FOLLOWER} and terms {}.
2020-02-25 13:44:26.773 [Broker-1-TopologyManager] [Broker-1-zb-actors-0] DEBUG io.zeebe.broker.clustering - Received metadata change for 0, partitions {3=FOLLOWER} terms {}
2020-02-25 13:44:27.023 [GatewayTopologyManager] [Broker-1-zb-actors-0] DEBUG io.zeebe.gateway - Received metadata change from Broker 0, partitions {2=FOLLOWER, 3=FOLLOWER} and terms {}.
2020-02-25 13:44:27.023 [Broker-1-TopologyManager] [Broker-1-zb-actors-0] DEBUG io.zeebe.broker.clustering - Received metadata change for 0, partitions {2=FOLLOWER, 3=FOLLOWER} terms {}


2020-02-25 13:50:22.051 [io.zeebe.gateway.impl.broker.BrokerRequestManager] [Broker-2-zb-actors-0] ERROR io.zeebe.gateway - Error handling gRPC request
io.grpc.StatusRuntimeException: INTERNAL: Unexpected error occurred during the request processing
	at io.grpc.Status.asRuntimeException(Status.java:524) ~[grpc-api-1.26.0.jar:1.26.0]
	at io.zeebe.gateway.EndpointManager.convertThrowable(EndpointManager.java:372) ~[zeebe-gateway-0.22.1.jar:0.22.1]
	at io.zeebe.gateway.impl.job.ActivateJobsHandler.logErrorResponse(ActivateJobsHandler.java:109) ~[zeebe-gateway-0.22.1.jar:0.22.1]
	at io.zeebe.gateway.impl.job.ActivateJobsHandler.lambda$activateJobs$1(ActivateJobsHandler.java:96) ~[zeebe-gateway-0.22.1.jar:0.22.1]
	at io.zeebe.gateway.impl.broker.BrokerRequestManager.lambda$sendRequest$3(BrokerRequestManager.java:147) ~[zeebe-gateway-0.22.1.jar:0.22.1]
	at io.zeebe.gateway.impl.broker.BrokerRequestManager.lambda$sendRequestInternal$5(BrokerRequestManager.java:183) ~[zeebe-gateway-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.future.FutureContinuationRunnable.run(FutureContinuationRunnable.java:33) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:76) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:115) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:107) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:91) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:195) [zeebe-util-0.22.1.jar:0.22.1]
Caused by: java.util.concurrent.TimeoutException: Request timed out after PT15S
	at io.zeebe.transport.impl.RequestContext.timeout(RequestContext.java:109) ~[zeebe-transport-0.22.1.jar:0.22.1]
	at io.zeebe.transport.impl.AtomixClientTransportAdapter.timeoutFuture(AtomixClientTransportAdapter.java:128) ~[zeebe-transport-0.22.1.jar:0.22.1]
	at io.zeebe.transport.impl.AtomixClientTransportAdapter.lambda$sendRequestWithRetry$0(AtomixClientTransportAdapter.java:60) ~[zeebe-transport-0.22.1.jar:0.22.1]
	... 6 more
2020-02-25 13:50:22.052 [io.zeebe.gateway.impl.broker.BrokerRequestManager] [Broker-2-zb-actors-0] WARN  io.zeebe.gateway - Failed to activate jobs for type http from partition 2
java.util.concurrent.TimeoutException: Request timed out after PT15S
	at io.zeebe.transport.impl.RequestContext.timeout(RequestContext.java:109) ~[zeebe-transport-0.22.1.jar:0.22.1]
	at io.zeebe.transport.impl.AtomixClientTransportAdapter.timeoutFuture(AtomixClientTransportAdapter.java:128) ~[zeebe-transport-0.22.1.jar:0.22.1]
	at io.zeebe.transport.impl.AtomixClientTransportAdapter.lambda$sendRequestWithRetry$0(AtomixClientTransportAdapter.java:60) ~[zeebe-transport-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:76) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:115) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:107) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:91) [zeebe-util-0.22.1.jar:0.22.1]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:195) [zeebe-util-0.22.1.jar:0.22.1]

```

</p>
</details>


**Environment:**
```
2020-02-25 13:53:58.999 [] [main] DEBUG io.zeebe.broker.system - Initializing configuration with base path /usr/local/zeebe
2020-02-25 13:53:59.065 [] [main] INFO  io.zeebe.broker.system - Version: 0.22.1
2020-02-25 13:53:58.459 [] [main] INFO  io.zeebe.broker.system - Starting broker 0 with configuration {
  "network": {
    "host": "beta-zeebe-zeebe-0.zeebe.beta.svc.cluster.local",
    "portOffset": 0,
    "maxMessageSize": "4M",
    "commandApi": {
      "defaultPort": 26501,
      "host": "beta-zeebe-zeebe-0.zeebe.beta.svc.cluster.local",
      "port": 26501,
      "advertisedHost": "beta-zeebe-zeebe-0.zeebe.beta.svc.cluster.local",
      "advertisedPort": 26501
    },
    "internalApi": {
      "defaultPort": 26502,
      "host": "beta-zeebe-zeebe-0.zeebe.beta.svc.cluster.local",
      "port": 26502,
      "advertisedHost": "beta-zeebe-zeebe-0.zeebe.beta.svc.cluster.local",
      "advertisedPort": 26502
    },
    "monitoringApi": {
      "defaultPort": 9600,
      "host": "beta-zeebe-zeebe-0.zeebe.beta.svc.cluster.local",
      "port": 9600,
      "advertisedHost": "beta-zeebe-zeebe-0.zeebe.beta.svc.cluster.local",
      "advertisedPort": 9600
    }
  },
  "cluster": {
    "initialContactPoints": [
      "beta-zeebe-zeebe-0.zeebe.beta.svc.cluster.local:26502",
      "beta-zeebe-zeebe-1.zeebe.beta.svc.cluster.local:26502",
      "beta-zeebe-zeebe-2.zeebe.beta.svc.cluster.local:26502"
    ],
    "partitionIds": [
      1,
      2,
      3
    ],
    "nodeId": 0,
    "partitionsCount": 3,
    "replicationFactor": 3,
    "clusterSize": 3,
    "clusterName": "zeebe-cluster",
    "gossipFailureTimeout": 10000,
    "gossipInterval": 250,
    "gossipProbeInterval": 1000
  },
  "threads": {
    "cpuThreadCount": 1,
    "ioThreadCount": 1
  },
  "data": {
    "directories": [
      "/usr/local/zeebe/data"
    ],
    "logSegmentSize": "512M",
    "snapshotPeriod": "15m",
    "raftSegmentSize": "512M",
    "maxSnapshots": 3
  },
  "exporters": [
    {
      "id": "elasticsearch",
      "className": "io.zeebe.exporter.ElasticsearchExporter",
      "args": {
        "index": {
          "workflowInstance": true,
          "messageSubscription": true,
          "prefix": "zeebe-beta",
          "jobBatch": true,
          "error": true,
          "message": true,
          "createTemplate": true,
          "command": true,
          "rejection": true,
          "workflowInstanceCreation": true,
          "variable": true,
          "variableDocument": true,
          "workflowInstanceSubscription": true,
          "event": true,
          "job": true,
          "incident": true,
          "deployment": true
        },
        "bulk": {
          "delay": 10.0,
          "size": 1000.0
        },
        "url": "my_url"
      }
    },
    {
      "id": "hazelcast",
      "jarPath": "/usr/local/zeebe/lib/zeebe-hazelcast-exporter.jar",
      "className": "io.zeebe.hazelcast.exporter.HazelcastExporter",
      "args": {
        "updatePosition": true,
        "port": 5701.0,
        "topicPrefix": "zeebe-",
        "enabledRecordTypes": "EVENT",
        "enabledValueTypes": "DEPLOYMENT,JOB,JOB_BATCH,WORKFLOW_INSTANCE,WORKFLOW_INSTANCE_CREATION,WORKFLOW_INSTANCE_SUBSCRIPTION,INCIDENT,TIMER,VARIABLE,MESSAGE,MESSAGE_SUBSCRIPTION,MESSAGE_START_EVENT_SUBSCRIPTION"
      }
    }
  ],
  "gateway": {
    "enable": true,
    "network": {
      "host": "0.0.0.0",
      "port": 26500,
      "minKeepAliveInterval": "30s"
    },
    "cluster": {
      "contactPoint": "beta-zeebe-zeebe-0.zeebe.beta.svc.cluster.local:26502",
      "maxMessageSize": "4M",
      "requestTimeout": "15s",
      "clusterName": "zeebe-cluster",
      "memberId": "gateway",
      "host": "0.0.0.0",
      "port": 26502
    },
    "threads": {
      "managementThreads": 1
    },
    "monitoring": {
      "enabled": false,
      "host": "beta-zeebe-zeebe-0.zeebe.beta.svc.cluster.local",
      "port": 9600
    },
    "security": {
      "enabled": false
    }
  },
  "backpressure": {
    "enabled": true,
    "useWindowed": true,
    "algorithm": "vegas"
  },
  "stepTimeout": "5m"
}
```

hmmm 2nd try didn't work, I can now confirm that the killed broker is not restarted (`chaos.log` shows broker 1 was killed, docker ps shows it's not running anymore):
```
❯ docker ps | grep broker
234b6c7cc947        camunda/zeebe:0.23.0-alpha1           "tini -- /usr/local/…"   51 minutes ago      Up 20 minutes       26500-26502/tcp          zeebe_broker_3
03e5004583d4        camunda/zeebe:0.23.0-alpha1           "tini -- /usr/local/…"   51 minutes ago      Up 20 minutes       26500-26502/tcp          zeebe_broker_2
```
relevant log from broker 1:
```
zeebe_broker_1 | 2020-02-28T13:03:51.031762800Z 2020-02-28 13:03:51.021 [Broker-0-StreamProcessor-1] [Broker-0-zb-actors-1] DEBUG io.zeebe.broker.workflow.repository - Pushed deployment 2251799813685334 to all partitions.
zeebe_broker_1 | 2020-02-28T13:03:51.266358600Z 2020-02-28 13:03:51.257 [Broker-0-StreamProcessor-1] [Broker-0-zb-actors-0] DEBUG io.zeebe.broker.workflow.repository - Deployment 2251799813685334 distributed to all partitions successfully.
zeebe_broker_1 | 2020-02-28T13:03:57.085443200Z 2020-02-28 13:03:57.072 [GatewayLongPollingJobHandler] [Broker-0-zb-actors-1] DEBUG io.zeebe.gateway - Remove blocking request BrokerActivateJobsRequest{requestDto={"type":"addAmount","worker":"magic-numbers-worker","timeout":10000,"maxJobsToActivate":2,"jobKeys":[],"jobs":[],"variables":[],"truncated":false}} for job type addAmount after timeout of PT20S
zeebe_broker_1 | 2020-02-28T13:03:57.209276400Z 2020-02-28 13:03:57.198 [GatewayLongPollingJobHandler] [Broker-0-zb-actors-0] DEBUG io.zeebe.gateway - Remove blocking request BrokerActivateJobsRequest{requestDto={"type":"outputNumber","worker":"magic-numbers-worker","timeout":10000,"maxJobsToActivate":2,"jobKeys":[],"jobs":[],"variables":[],"truncated":false}} for job type outputNumber after timeout of PT20S
zeebe_broker_1 | 2020-02-28T13:03:57.211730100Z 2020-02-28 13:03:57.200 [GatewayLongPollingJobHandler] [Broker-0-zb-actors-0] DEBUG io.zeebe.gateway - Remove blocking request BrokerActivateJobsRequest{requestDto={"type":"generateNumber","worker":"magic-numbers-worker","timeout":10000,"maxJobsToActivate":2,"jobKeys":[],"jobs":[],"variables":[],"truncated":false}} for job type generateNumber after timeout of PT20S
zeebe_broker_1 | 2020-02-28T13:04:06.147629800Z 2020-02-28 13:04:06.069 [io.zeebe.gateway.impl.broker.BrokerRequestManager] [Broker-0-zb-actors-1] ERROR io.zeebe.gateway - Error handling gRPC request
zeebe_broker_1 | 2020-02-28T13:04:06.155806100Z io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: Time out between gateway and broker: Request type command-api-2 timed out in 14999 milliseconds
zeebe_broker_1 | 2020-02-28T13:04:06.155852000Z 	at io.grpc.Status.asRuntimeException(Status.java:524) ~[grpc-api-1.27.0.jar:1.27.0]
zeebe_broker_1 | 2020-02-28T13:04:06.156362000Z 	at io.zeebe.gateway.EndpointManager.convertThrowable(EndpointManager.java:382) ~[zeebe-gateway-0.23.0-alpha1.jar:0.23.0-alpha1]
zeebe_broker_1 | 2020-02-28T13:04:06.156394900Z 	at io.zeebe.gateway.EndpointManager.lambda$sendRequest$3(EndpointManager.java:307) ~[zeebe-gateway-0.23.0-alpha1.jar:0.23.0-alpha1]
zeebe_broker_1 | 2020-02-28T13:04:06.156427400Z 	at io.zeebe.gateway.impl.broker.BrokerRequestManager.lambda$sendRequest$3(BrokerRequestManager.java:147) ~[zeebe-gateway-0.23.0-alpha1.jar:0.23.0-alpha1]
zeebe_broker_1 | 2020-02-28T13:04:06.156443800Z 	at io.zeebe.gateway.impl.broker.BrokerRequestManager.lambda$sendRequestInternal$5(BrokerRequestManager.java:183) ~[zeebe-gateway-0.23.0-alpha1.jar:0.23.0-alpha1]
zeebe_broker_1 | 2020-02-28T13:04:06.156476900Z 	at io.zeebe.util.sched.future.FutureContinuationRunnable.run(FutureContinuationRunnable.java:33) [zeebe-util-0.23.0-alpha1.jar:0.23.0-alpha1]
zeebe_broker_1 | 2020-02-28T13:04:06.156497900Z 	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:76) [zeebe-util-0.23.0-alpha1.jar:0.23.0-alpha1]
zeebe_broker_1 | 2020-02-28T13:04:06.156529000Z 	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-0.23.0-alpha1.jar:0.23.0-alpha1]
zeebe_broker_1 | 2020-02-28T13:04:06.156548400Z 	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:115) [zeebe-util-0.23.0-alpha1.jar:0.23.0-alpha1]
zeebe_broker_1 | 2020-02-28T13:04:06.156576700Z 	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:107) [zeebe-util-0.23.0-alpha1.jar:0.23.0-alpha1]
zeebe_broker_1 | 2020-02-28T13:04:06.156602500Z 	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:91) [zeebe-util-0.23.0-alpha1.jar:0.23.0-alpha1]
zeebe_broker_1 | 2020-02-28T13:04:06.156622900Z 	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:195) [zeebe-util-0.23.0-alpha1.jar:0.23.0-alpha1]
zeebe_broker_1 | 2020-02-28T13:04:06.156643600Z Caused by: java.util.concurrent.TimeoutException: Request type command-api-2 timed out in 14999 milliseconds
zeebe_broker_1 | 2020-02-28T13:04:06.156661500Z 	at io.atomix.cluster.messaging.impl.AbstractClientConnection$Callback.timeout(AbstractClientConnection.java:159) ~[atomix-cluster-3.2.0-alpha11.jar:?]
zeebe_broker_1 | 2020-02-28T13:04:06.156685400Z 	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) ~[?:?]
zeebe_broker_1 | 2020-02-28T13:04:06.156706800Z 	at java.util.concurrent.FutureTask.run(Unknown Source) ~[?:?]
zeebe_broker_1 | 2020-02-28T13:04:06.156727300Z 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) ~[?:?]
zeebe_broker_1 | 2020-02-28T13:04:06.156747700Z 	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[?:?]
zeebe_broker_1 | 2020-02-28T13:04:06.156781200Z 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[?:?]
zeebe_broker_1 | 2020-02-28T13:04:06.156802500Z 	at java.lang.Thread.run(Unknown Source) ~[?:?]
zeebe_broker_1 | 2020-02-28T13:04:10.842414300Z 2020-02-28 13:04:10.830 [] [Broker close Thread] INFO  io.zeebe.broker.system - Closing Broker-0 [1/11]: zeebe partitions
zeebe_broker_1 | 2020-02-28T13:04:11.515362400Z 2020-02-28 13:04:11,501 Broker-0-zb-fs-workers-1 WARN Unable to register Log4j shutdown hook because JVM is shutting down. Using SimpleLogger
```

--- a/broker/src/main/java/io/zeebe/broker/engine/impl/SubscriptionApiCommandMessageHandlerService.java
+++ b/broker/src/main/java/io/zeebe/broker/engine/impl/SubscriptionApiCommandMessageHandlerService.java
@@ -37,15 +37,16 @@ public final class SubscriptionApiCommandMessageHandlerService extends Actor
   }
 
   @Override
+  protected void onActorStarting() {
+    final SubscriptionCommandMessageHandler messageHandler =
+        new SubscriptionCommandMessageHandler(actor::call, leaderPartitions::get);
+    atomix.getCommunicationService().subscribe("subscription", messageHandler);
+  }
+
+  @Override
   public void onBecomingFollower(
       final int partitionId, final long term, final LogStream logStream) {
-    actor.submit(
-        () -> {
-          final var recordWriter = leaderPartitions.remove(partitionId);
-          if (recordWriter != null) {
-            recordWriter.close();
-          }
-        });
+    actor.submit(() -> leaderPartitions.remove(partitionId));
   }
 
   @Override
@@ -67,12 +68,5 @@ public final class SubscriptionApiCommandMessageHandlerService extends Actor
                     }
                   });
         });
-  }
-
-  @Override
-  protected void onActorStarting() {
-    final SubscriptionCommandMessageHandler messageHandler =
-        new SubscriptionCommandMessageHandler(actor::call, leaderPartitions::get);
-    atomix.getCommunicationService().subscribe("subscription", messageHandler);
   }
 }

--- a/broker/src/main/java/io/zeebe/broker/system/management/LeaderManagementRequestHandler.java
+++ b/broker/src/main/java/io/zeebe/broker/system/management/LeaderManagementRequestHandler.java
@@ -33,14 +33,7 @@ public final class LeaderManagementRequestHandler extends Actor implements Parti
   @Override
   public void onBecomingFollower(
       final int partitionId, final long term, final LogStream logStream) {
-    actor.submit(
-        () -> {
-          final var recordWriter = leaderForPartitions.remove(partitionId);
-
-          if (recordWriter != null) {
-            recordWriter.close();
-          }
-        });
+    actor.submit(() -> leaderForPartitions.remove(partitionId));
   }
 
   @Override

--- a/broker/src/main/java/io/zeebe/broker/system/partitions/ZeebePartition.java
+++ b/broker/src/main/java/io/zeebe/broker/system/partitions/ZeebePartition.java
@@ -98,6 +98,8 @@ public final class ZeebePartition extends Actor
   private final HealthMonitor criticalComponentsHealthMonitor;
   private final ZeebeIndexMapping zeebeIndexMapping;
   private final HealthMetrics healthMetrics;
+  private AtomixLogStorage atomixLogStorage;
+  private long deferredCommitPosition;
 
   public ZeebePartition(
       final BrokerInfo localBroker,
@@ -287,6 +289,35 @@ public final class ZeebePartition extends Actor
   }
 
   private ActorFuture<Void> basePartitionInstallation() {
+    final ActorFuture installFuture = new CompletableActorFuture();
+    openLogStream()
+        .onComplete(
+            (log, error) -> {
+              if (error == null) {
+                this.logStream = log;
+                if (deferredCommitPosition > 0) {
+                  logStream.setCommitPosition(deferredCommitPosition);
+                  deferredCommitPosition = -1;
+                }
+                installStorageServices()
+                    .onComplete(
+                        (deletionService, errorInstall) -> {
+                          if (errorInstall == null) {
+                            installFuture.complete(deletionService);
+                          } else {
+                            installFuture.completeExceptionally(errorInstall);
+                          }
+                        });
+              } else {
+                LOG.error("Failed to install log stream for partition {}", partitionId, error);
+                installFuture.completeExceptionally(error);
+                onFailure();
+              }
+            });
+    return installFuture;
+  }
+
+  private ActorFuture<Void> installStorageServices() {
     snapshotStorage = createSnapshotStorage();
     snapshotController = createSnapshotController();
 
@@ -322,7 +353,6 @@ public final class ZeebePartition extends Actor
   }
 
   private SnapshotStorage createSnapshotStorage() {
-
     final var reader =
         new AtomixLogStorageReader(
             zeebeIndexMapping, atomixRaftPartition.getServer().openReader(-1, Mode.COMMITS));
@@ -432,8 +462,7 @@ public final class ZeebePartition extends Actor
     closeActorsFuture.onComplete(
         (v, t) -> {
           if (t == null) {
-            tearDownBaseInstallation();
-            closingPartitionFuture.complete(null);
+            tearDownBaseInstallation(closingPartitionFuture);
           } else {
             closingPartitionFuture.completeExceptionally(t);
           }
@@ -442,16 +471,32 @@ public final class ZeebePartition extends Actor
     return closingPartitionFuture;
   }
 
-  private void tearDownBaseInstallation() {
-    if (closeStateReplication()) {
-      return;
-    }
-
-    if (closeSnapshotController()) {
-      return;
-    }
-
+  private void tearDownBaseInstallation(final CompletableActorFuture<Void> closeFuture) {
+    closeStateReplication();
+    closeSnapshotController();
     closeSnapshotStorage();
+
+    closeLogStream()
+        .onComplete(
+            (closed, error) -> {
+              if (error == null) {
+                closeFuture.complete(null);
+              } else {
+                LOG.error(
+                    "Unexpected error on closing logstream for partition {}", partitionId, error);
+                closeFuture.completeExceptionally(error);
+              }
+            });
+  }
+
+  private ActorFuture<Void> closeLogStream() {
+    if (logStream == null) {
+      return CompletableActorFuture.completed(null);
+    }
+
+    final LogStream logStreamToClose = logStream;
+    logStream = null;
+    return logStreamToClose.closeAsync();
   }
 
   private void closeSnapshotStorage() {
@@ -469,9 +514,9 @@ public final class ZeebePartition extends Actor
     }
   }
 
-  private boolean closeSnapshotController() {
+  private void closeSnapshotController() {
     if (snapshotController == null) {
-      return true;
+      return;
     }
 
     try {
@@ -486,12 +531,11 @@ public final class ZeebePartition extends Actor
       snapshotController = null;
       zeebeDb = null;
     }
-    return false;
   }
 
-  private boolean closeStateReplication() {
+  private void closeStateReplication() {
     if (stateReplication == null) {
-      return true;
+      return;
     }
 
     try {
@@ -501,7 +545,6 @@ public final class ZeebePartition extends Actor
     } finally {
       stateReplication = null;
     }
-    return false;
   }
 
   private void stepByStepClosing(
@@ -530,7 +573,15 @@ public final class ZeebePartition extends Actor
   @Override
   public <T extends RaftLogEntry> void onCommit(final Indexed<T> indexed) {
     if (indexed.type() == ZeebeEntry.class) {
-      this.logStream.setCommitPosition(indexed.<ZeebeEntry>cast().entry().highestPosition());
+      actor.run(
+          () -> {
+            final long commitPosition = indexed.<ZeebeEntry>cast().entry().highestPosition();
+            if (this.logStream == null) {
+              this.deferredCommitPosition = commitPosition;
+              return;
+            }
+            this.logStream.setCommitPosition(commitPosition);
+          });
     }
   }
 
@@ -541,34 +592,11 @@ public final class ZeebePartition extends Actor
 
   @Override
   public void onActorStarting() {
-    final var atomixLogStorage =
-        AtomixLogStorage.ofPartition(zeebeIndexMapping, atomixRaftPartition);
-
-    LogStream.builder()
-        .withLogStorage(atomixLogStorage)
-        .withLogName(atomixRaftPartition.name())
-        .withNodeId(localBroker.getNodeId())
-        .withPartitionId(atomixRaftPartition.id().id())
-        .withMaxFragmentSize(maxFragmentSize)
-        .withActorScheduler(scheduler)
-        .buildAsync()
-        .onComplete(
-            (log, error) -> {
-              if (error == null) {
-                this.logStream = log;
-                atomixRaftPartition.getServer().addCommitListener(this);
-                atomixRaftPartition.addRoleChangeListener(this);
-                onRecovered();
-                onRoleChange(atomixRaftPartition.getRole(), atomixRaftPartition.term());
-              } else {
-                LOG.error(
-                    "Failed to install log stream service for partition {}",
-                    atomixRaftPartition.id().id(),
-                    error);
-                actor.close();
-                onFailure();
-              }
-            });
+    atomixLogStorage = AtomixLogStorage.ofPartition(zeebeIndexMapping, atomixRaftPartition);
+    atomixRaftPartition.getServer().addCommitListener(this);
+    atomixRaftPartition.addRoleChangeListener(this);
+    onRoleChange(atomixRaftPartition.getRole(), atomixRaftPartition.term());
+    onRecovered();
   }
 
   @Override
@@ -589,18 +617,26 @@ public final class ZeebePartition extends Actor
                       atomixRaftPartition.removeRoleChangeListener(this);
                       atomixRaftPartition.getServer().removeCommitListener(this);
 
-                      final ActorFuture<Void> logStreamCloseFuture = logStream.closeAsync();
                       if (t == null) {
-                        logStreamCloseFuture.onComplete(closeFuture);
+                        closeFuture.complete(null);
                       } else {
-                        // we want to close the log stream any way
-                        logStreamCloseFuture.onComplete(
-                            (v2, t2) -> closeFuture.completeExceptionally(t));
+                        closeFuture.completeExceptionally(t);
                       }
                     }));
     closeFuture.join();
 
     super.close();
+  }
+
+  private ActorFuture<LogStream> openLogStream() {
+    return LogStream.builder()
+        .withLogStorage(atomixLogStorage)
+        .withLogName(atomixRaftPartition.name())
+        .withNodeId(localBroker.getNodeId())
+        .withPartitionId(atomixRaftPartition.id().id())
+        .withMaxFragmentSize(maxFragmentSize)
+        .withActorScheduler(scheduler)
+        .buildAsync();
   }
 
   @Override

--- a/broker/src/main/java/io/zeebe/broker/transport/commandapi/CommandApiRequestHandler.java
+++ b/broker/src/main/java/io/zeebe/broker/transport/commandapi/CommandApiRequestHandler.java
@@ -197,11 +197,7 @@ final class CommandApiRequestHandler implements RequestHandler {
   void removePartition(final LogStream logStream) {
     cmdQueue.add(
         () -> {
-          final var recordWriter = leadingStreams.remove(logStream.getPartitionId());
-          if (recordWriter != null) {
-            recordWriter.close();
-          }
-
+          leadingStreams.remove(logStream.getPartitionId());
           partitionLimiters.remove(logStream.getPartitionId());
         });
   }

--- a/engine/src/main/java/io/zeebe/engine/processor/NoopTypedStreamWriter.java
+++ b/engine/src/main/java/io/zeebe/engine/processor/NoopTypedStreamWriter.java
@@ -85,9 +85,4 @@ final class NoopTypedStreamWriter implements TypedStreamWriter {
   public long flush() {
     return 0;
   }
-
-  @Override
-  public void close() {
-    // no op implementation
-  }
 }

--- a/engine/src/main/java/io/zeebe/engine/processor/StreamProcessor.java
+++ b/engine/src/main/java/io/zeebe/engine/processor/StreamProcessor.java
@@ -153,7 +153,6 @@ public class StreamProcessor extends Actor {
   @Override
   protected void onActorClosing() {
     processingContext.getLogStreamReader().close();
-    processingContext.getLogStreamWriter().close();
 
     if (onCommitPositionUpdatedCondition != null) {
       logStream.removeOnCommitPositionUpdatedCondition(onCommitPositionUpdatedCondition);

--- a/engine/src/main/java/io/zeebe/engine/processor/TypedCommandWriter.java
+++ b/engine/src/main/java/io/zeebe/engine/processor/TypedCommandWriter.java
@@ -10,11 +10,10 @@ package io.zeebe.engine.processor;
 import io.zeebe.msgpack.UnpackedObject;
 import io.zeebe.protocol.impl.record.RecordMetadata;
 import io.zeebe.protocol.record.intent.Intent;
-import io.zeebe.util.CloseableSilently;
 import java.util.function.Consumer;
 
 /** Things that any actor can write to a partition. */
-public interface TypedCommandWriter extends CloseableSilently {
+public interface TypedCommandWriter {
 
   void appendNewCommand(Intent intent, UnpackedObject value);
 

--- a/engine/src/main/java/io/zeebe/engine/processor/TypedCommandWriterImpl.java
+++ b/engine/src/main/java/io/zeebe/engine/processor/TypedCommandWriterImpl.java
@@ -120,9 +120,4 @@ public class TypedCommandWriterImpl implements TypedCommandWriter {
   public long flush() {
     return batchWriter.tryWrite();
   }
-
-  @Override
-  public void close() {
-    batchWriter.close();
-  }
 }

--- a/engine/src/test/java/io/zeebe/engine/util/TestStreams.java
+++ b/engine/src/test/java/io/zeebe/engine/util/TestStreams.java
@@ -371,7 +371,6 @@ public final class TestStreams {
 
     @Override
     public void close() {
-      logStreamWriter.close();
       logStream.close();
       logStorageRule.close();
     }

--- a/logstreams/src/main/java/io/zeebe/logstreams/impl/log/LogStreamBatchWriterImpl.java
+++ b/logstreams/src/main/java/io/zeebe/logstreams/impl/log/LogStreamBatchWriterImpl.java
@@ -60,13 +60,10 @@ public final class LogStreamBatchWriterImpl implements LogStreamBatchWriter, Log
 
   private BufferWriter metadataWriter;
   private BufferWriter valueWriter;
-  private final Runnable closeCallback;
 
-  LogStreamBatchWriterImpl(
-      final int partitionId, final Dispatcher dispatcher, final Runnable closeCallback) {
+  LogStreamBatchWriterImpl(final int partitionId, final Dispatcher dispatcher) {
     this.logWriteBuffer = dispatcher;
     this.logId = partitionId;
-    this.closeCallback = closeCallback;
 
     reset();
   }
@@ -293,10 +290,5 @@ public final class LogStreamBatchWriterImpl implements LogStreamBatchWriter, Log
 
     bufferWriterInstance.reset();
     metadataWriterInstance.reset();
-  }
-
-  @Override
-  public void close() {
-    closeCallback.run();
   }
 }

--- a/logstreams/src/main/java/io/zeebe/logstreams/impl/log/LogStreamImpl.java
+++ b/logstreams/src/main/java/io/zeebe/logstreams/impl/log/LogStreamImpl.java
@@ -27,7 +27,6 @@ import java.io.IOException;
 import java.io.UncheckedIOException;
 import java.util.ArrayList;
 import java.util.List;
-import java.util.concurrent.atomic.AtomicInteger;
 import java.util.function.BiConsumer;
 import org.slf4j.Logger;
 
@@ -53,8 +52,6 @@ public final class LogStreamImpl extends Actor implements LogStream, AutoCloseab
   private long commitPosition;
   private Throwable closeError; // set if any error occurred during closeAsync
   private final String actorName;
-
-  private final AtomicInteger openWriterCount = new AtomicInteger(0);
 
   LogStreamImpl(
       final ActorScheduler actorScheduler,
@@ -101,50 +98,6 @@ public final class LogStreamImpl extends Actor implements LogStream, AutoCloseab
   }
 
   @Override
-  public String getName() {
-    return actorName;
-  }
-
-  @Override
-  public void close() {
-    closeAsync().join();
-  }
-
-  @Override
-  protected void onActorClosing() {
-    LOG.info("On closing logstream {} close {} readers", logName, readers.size());
-    readers.forEach(LogStreamReader::close);
-    LOG.info("Close log storage with name {}", logName);
-    logStorage.close();
-  }
-
-  @Override
-  protected void onActorClosed() {
-    if (closeError != null) {
-      closeFuture.completeExceptionally(closeError);
-    } else {
-      closeFuture.complete(null);
-    }
-  }
-
-  @Override
-  public ActorFuture<Void> closeAsync() {
-    if (actor.isClosed()) {
-      return closeFuture;
-    }
-
-    actor.call(
-        () ->
-            closeAppender()
-                .onComplete(
-                    (nothing, appenderError) -> {
-                      closeError = appenderError;
-                      actor.close();
-                    }));
-    return closeFuture;
-  }
-
-  @Override
   public ActorFuture<Long> getCommitPositionAsync() {
     return actor.call(() -> commitPosition);
   }
@@ -152,22 +105,6 @@ public final class LogStreamImpl extends Actor implements LogStream, AutoCloseab
   @Override
   public void setCommitPosition(final long commitPosition) {
     actor.call(() -> internalSetCommitPosition(commitPosition));
-  }
-
-  private void internalSetCommitPosition(final long commitPosition) {
-    this.commitPosition = commitPosition;
-
-    onCommitPositionUpdatedConditions.signalConsumers();
-  }
-
-  @Override
-  public void registerOnCommitPositionUpdatedCondition(final ActorCondition condition) {
-    actor.call(() -> onCommitPositionUpdatedConditions.registerConsumer(condition));
-  }
-
-  @Override
-  public void removeOnCommitPositionUpdatedCondition(final ActorCondition condition) {
-    actor.call(() -> onCommitPositionUpdatedConditions.removeConsumer(condition));
   }
 
   @Override
@@ -204,21 +141,75 @@ public final class LogStreamImpl extends Actor implements LogStream, AutoCloseab
     return writerFuture;
   }
 
+  @Override
+  public void registerOnCommitPositionUpdatedCondition(final ActorCondition condition) {
+    actor.call(() -> onCommitPositionUpdatedConditions.registerConsumer(condition));
+  }
+
+  @Override
+  public void removeOnCommitPositionUpdatedCondition(final ActorCondition condition) {
+    actor.call(() -> onCommitPositionUpdatedConditions.removeConsumer(condition));
+  }
+
+  @Override
+  public String getName() {
+    return actorName;
+  }
+
+  @Override
+  protected void onActorClosing() {
+    LOG.info("On closing logstream {} close {} readers", logName, readers.size());
+    readers.forEach(LogStreamReader::close);
+    LOG.info("Close log storage with name {}", logName);
+    logStorage.close();
+  }
+
+  @Override
+  protected void onActorClosed() {
+    if (closeError != null) {
+      closeFuture.completeExceptionally(closeError);
+    } else {
+      closeFuture.complete(null);
+    }
+  }
+
+  @Override
+  public void close() {
+    closeAsync().join();
+  }
+
+  @Override
+  public ActorFuture<Void> closeAsync() {
+    if (actor.isClosed()) {
+      return closeFuture;
+    }
+
+    actor.call(
+        () ->
+            closeAppender()
+                .onComplete(
+                    (nothing, appenderError) -> {
+                      closeError = appenderError;
+                      actor.close();
+                    }));
+    return closeFuture;
+  }
+
+  private void internalSetCommitPosition(final long commitPosition) {
+    if (commitPosition > this.commitPosition) {
+      this.commitPosition = commitPosition;
+      onCommitPositionUpdatedConditions.signalConsumers();
+    }
+  }
+
   private <T extends LogStreamWriter> void createWriter(
       final CompletableActorFuture<T> writerFuture, final WriterCreator<T> creator) {
-    final var alreadyOpenWriters = openWriterCount.getAndIncrement();
-    if (alreadyOpenWriters == 0) {
-      openAppender().onComplete(onOpenAppender(writerFuture, creator));
-    } else if (appender != null) {
-      writerFuture.complete(creator.create(partitionId, writeBuffer, this::releaseWriter));
+    if (appender != null) {
+      writerFuture.complete(creator.create(partitionId, writeBuffer));
     } else if (appenderFuture != null) {
       appenderFuture.onComplete(onOpenAppender(writerFuture, creator));
     } else {
-      final var errorMsg =
-          String.format(
-              "Expected to have an open appender, since we have already %d open writers",
-              alreadyOpenWriters);
-      writerFuture.completeExceptionally(new IllegalStateException(errorMsg));
+      openAppender().onComplete(onOpenAppender(writerFuture, creator));
     }
   }
 
@@ -226,21 +217,11 @@ public final class LogStreamImpl extends Actor implements LogStream, AutoCloseab
       final CompletableActorFuture<T> writerFuture, final WriterCreator<T> creator) {
     return (openedAppender, errorOnOpeningAppender) -> {
       if (errorOnOpeningAppender == null) {
-        writerFuture.complete(creator.create(partitionId, writeBuffer, this::releaseWriter));
+        writerFuture.complete(creator.create(partitionId, writeBuffer));
       } else {
         writerFuture.completeExceptionally(errorOnOpeningAppender);
       }
     };
-  }
-
-  private void releaseWriter() {
-    actor.run(
-        () -> {
-          final var remainingOpenWriters = openWriterCount.decrementAndGet();
-          if (remainingOpenWriters == 0) {
-            closeAppender();
-          }
-        });
   }
 
   private ActorFuture<Void> closeAppender() {
@@ -336,6 +317,6 @@ public final class LogStreamImpl extends Actor implements LogStream, AutoCloseab
 
   @FunctionalInterface
   private interface WriterCreator<T extends LogStreamWriter> {
-    T create(int partitionId, Dispatcher dispatcher, Runnable closeCallback);
+    T create(int partitionId, Dispatcher dispatcher);
   }
 }

--- a/logstreams/src/main/java/io/zeebe/logstreams/impl/log/LogStreamWriterImpl.java
+++ b/logstreams/src/main/java/io/zeebe/logstreams/impl/log/LogStreamWriterImpl.java
@@ -39,13 +39,10 @@ public final class LogStreamWriterImpl implements LogStreamRecordWriter {
   private long sourceRecordPosition = -1L;
   private BufferWriter metadataWriter;
   private BufferWriter valueWriter;
-  private final Runnable closeCallback;
 
-  LogStreamWriterImpl(
-      final int partitionId, final Dispatcher logWriteBuffer, final Runnable closeCallback) {
+  LogStreamWriterImpl(final int partitionId, final Dispatcher logWriteBuffer) {
     this.logWriteBuffer = logWriteBuffer;
     this.partitionId = partitionId;
-    this.closeCallback = closeCallback;
 
     reset();
   }
@@ -170,10 +167,5 @@ public final class LogStreamWriterImpl implements LogStreamRecordWriter {
     } while (claimedPosition == RESULT_PADDING_AT_END_OF_PARTITION);
 
     return claimedPosition - DataFrameDescriptor.alignedFramedLength(framedLength);
-  }
-
-  @Override
-  public void close() {
-    closeCallback.run();
   }
 }

--- a/logstreams/src/main/java/io/zeebe/logstreams/log/LogStreamWriter.java
+++ b/logstreams/src/main/java/io/zeebe/logstreams/log/LogStreamWriter.java
@@ -7,9 +7,7 @@
  */
 package io.zeebe.logstreams.log;
 
-import io.zeebe.util.CloseableSilently;
-
-public interface LogStreamWriter extends CloseableSilently {
+public interface LogStreamWriter {
 
   /**
    * Attempts to write the event to the underlying stream.

--- a/logstreams/src/test/java/io/zeebe/logstreams/impl/log/LogStorageAppenderTest.java
+++ b/logstreams/src/test/java/io/zeebe/logstreams/impl/log/LogStorageAppenderTest.java
@@ -72,7 +72,7 @@ public final class LogStorageAppenderTest {
     appender =
         new LogStorageAppender(
             "appender", PARTITION_ID, logStorage, subscription, MAX_FRAGMENT_SIZE);
-    writer = new LogStreamWriterImpl(PARTITION_ID, dispatcher, () -> {});
+    writer = new LogStreamWriterImpl(PARTITION_ID, dispatcher);
     reader = new LogStreamReaderImpl(logStorage);
   }
 

--- a/logstreams/src/test/java/io/zeebe/logstreams/log/LogStreamBatchWriterTest.java
+++ b/logstreams/src/test/java/io/zeebe/logstreams/log/LogStreamBatchWriterTest.java
@@ -414,4 +414,16 @@ public final class LogStreamBatchWriterTest {
     // then
     assertThat(pos).isEqualTo(0);
   }
+
+  @Test
+  public void shouldFailToWriteOnClosedLogStream() {
+    // given
+    logStreamRule.getLogStream().close();
+
+    // when
+    final long pos = writer.event().key(1).value(EVENT_VALUE_1).done().tryWrite();
+
+    // then
+    assertThat(pos).isEqualTo(-1);
+  }
 }

--- a/logstreams/src/test/java/io/zeebe/logstreams/log/LogStreamTest.java
+++ b/logstreams/src/test/java/io/zeebe/logstreams/log/LogStreamTest.java
@@ -92,6 +92,25 @@ public final class LogStreamTest {
     assertThatThrownBy(() -> logStream.newLogStreamBatchWriter()).hasMessage("Actor is closed");
   }
 
+  @Test
+  public void shouldIncreasePositionOnRestart() {
+    // given
+    final LogStreamRecordWriter writer = logStream.newLogStreamRecordWriter();
+    writer.value(wrapString("value")).tryWrite();
+    writer.value(wrapString("value")).tryWrite();
+    writer.value(wrapString("value")).tryWrite();
+    final long positionBeforeClose = writer.value(wrapString("value")).tryWrite();
+
+    // when
+    logStream.close();
+    logStreamRule.createLogStream();
+    final LogStreamRecordWriter newWriter = logStreamRule.getLogStream().newLogStreamRecordWriter();
+    final long positionAfterReOpen = newWriter.value(wrapString("value")).tryWrite();
+
+    // then
+    assertThat(positionAfterReOpen).isGreaterThan(positionBeforeClose);
+  }
+
   static long writeEvent(final SynchronousLogStream logStream) {
     return writeEvent(logStream, wrapString("event"));
   }

--- a/logstreams/src/test/java/io/zeebe/logstreams/log/LogStreamWriterTest.java
+++ b/logstreams/src/test/java/io/zeebe/logstreams/log/LogStreamWriterTest.java
@@ -62,7 +62,6 @@ public final class LogStreamWriterTest {
 
   @After
   public void tearDown() {
-    writer.close();
     writer = null;
   }
 
@@ -195,8 +194,6 @@ public final class LogStreamWriterTest {
     final long firstPosition = writer.key(123L).value(EVENT_VALUE).tryWrite();
 
     // when
-    writer.close();
-
     final SynchronousLogStream logStream = logStreamRule.getLogStream();
     writer = logStream.newLogStreamRecordWriter();
     final long secondPosition = writer.key(124L).value(EVENT_VALUE).tryWrite();
@@ -214,7 +211,6 @@ public final class LogStreamWriterTest {
     writerRule.waitForPositionToBeAppended(firstPosition);
 
     // when
-    writer.close();
     writerRule.closeWriter();
 
     final SynchronousLogStream logStream = logStreamRule.getLogStream();

--- a/logstreams/src/test/java/io/zeebe/logstreams/util/LogStreamRule.java
+++ b/logstreams/src/test/java/io/zeebe/logstreams/util/LogStreamRule.java
@@ -69,7 +69,7 @@ public final class LogStreamRule extends ExternalResource {
       final UnaryOperator<Builder> builder) {
     logStorageRule = new AtomixLogStorageRule(temporaryFolder);
     this.logStorageRule.open(builder);
-    startLogStream();
+    createLogStream();
     return logStream;
   }
 
@@ -79,7 +79,7 @@ public final class LogStreamRule extends ExternalResource {
     actorSchedulerRule.before();
 
     if (shouldStartByDefault) {
-      startLogStream();
+      createLogStream();
     }
   }
 
@@ -90,7 +90,7 @@ public final class LogStreamRule extends ExternalResource {
     actorSchedulerRule.after();
   }
 
-  private void startLogStream() {
+  public void createLogStream() {
     final ActorScheduler actorScheduler = actorSchedulerRule.get();
 
     if (logStorageRule == null) {

--- a/logstreams/src/test/java/io/zeebe/logstreams/util/LogStreamWriterRule.java
+++ b/logstreams/src/test/java/io/zeebe/logstreams/util/LogStreamWriterRule.java
@@ -36,10 +36,7 @@ public final class LogStreamWriterRule extends ExternalResource {
   }
 
   public void closeWriter() {
-    if (logStreamWriter != null) {
-      logStreamWriter.close();
-      logStreamWriter = null;
-    }
+    logStreamWriter = null;
   }
 
   public long writeEvents(final int count, final DirectBuffer event) {

